src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveResourceService.cs

```diff
@@ -119,6 +121,7 @@ public async Task<RetrieveResourceResponse> GetInstanceResourceAsync(RetrieveRes
                             isOriginalTransferSyntaxRequested,
                             requestedTransferSyntax);
 
+                        _dicomRequestContextAccessor.RequestContext.PartCount = message.Frames.Count();
```

Comment:
nit: can the `Frames` property be some sort of read-only collection instead of only `IEnumerable<T>`?